File: dotnet/src/Microsoft.Agents/ChatCompletion/ChatClientAgentThread.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.AI;
using Microsoft.Shared.Diagnostics;

namespace Microsoft.Agents;

/// <summary>
/// Chat client agent thread.
/// </summary>
public class ChatClientAgentThread : AgentThread
{
    private readonly List<ChatMessage> _chatMessages = [];

    /// <summary>
    /// Initializes a new instance of the <see cref="ChatClientAgentThread"/> class.
    /// </summary>
    public ChatClientAgentThread()
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ChatClientAgentThread"/> class.
    /// </summary>
    /// <param name="id">The id of an existing server side thread to continue.</param>
    /// <remarks>
    /// This constructor creates a <see cref="ChatClientAgentThread"/> that supports in-service message storage.
    /// </remarks>
    public ChatClientAgentThread(string id)
    {
        Throw.IfNullOrWhitespace(id);

        this.Id = id;
        this.StorageLocation = ChatClientAgentThreadType.ConversationId;
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ChatClientAgentThread"/> class.
    /// </summary>
    /// <param name="messages">A set of initial messages to seed the thread with.</param>
    /// <remarks>
    /// This constructor creates a <see cref="ChatClientAgentThread"/> that supports local in-memory message storage.
    /// </remarks>
    public ChatClientAgentThread(IEnumerable<ChatMessage> messages)
    {
        Throw.IfNull(messages);

        this._chatMessages.AddRange(messages);
        this.StorageLocation = ChatClientAgentThreadType.AgentThreadManaged;
    }

    /// <summary>
    /// Gets the location of the thread contents.
    /// </summary>
    internal virtual ChatClientAgentThreadType? StorageLocation { get; set; }

#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
    /// <inheritdoc/>
    internal virtual async IAsyncEnumerable<ChatMessage> GetMessagesAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        foreach (var message in this._chatMessages)
        {
            yield return message;
        }
    }

#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously

    /// <summary>
    /// This method is called when new messages have been contributed to the chat by any participant.
    /// </summary>
    /// <remarks>
    /// Inheritors can use this method to update their context based on the new message.
    /// </remarks>
    /// <param name="newMessages">The new messages.</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
    /// <returns>A task that completes when the context has been updated.</returns>
    internal virtual Task OnNewMessagesAsync(IReadOnlyCollection<ChatMessage> newMessages, CancellationToken cancellationToken = default)
    {
        switch (this.StorageLocation)
        {
            case ChatClientAgentThreadType.AgentThreadManaged:
                this._chatMessages.AddRange(newMessages);
                break;
            case ChatClientAgentThreadType.ConversationId:
                // If the thread messages are stored in the service
                // there is nothing to do here, since invoking the
                // service should already update the thread.
                break;
            default:
                throw new UnreachableException();
        }

        return Task.CompletedTask;
    }
}

File: python/azure-ai-multi-agent/ai/agent/chat_completion_agent/chat_client_agent_thread.py
# ---------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# ---------------------------------------------------------

import asyncio
from typing import AsyncIterator, List, Optional, Dict, Any

from ..abstract_agent.agent_thread import AgentThread
from ..abstract_agent.messages_retrievable_thread import MessagesRetrievableThread
from ..common import ChatMessage

class ChatClientAgentThreadType:
    """
    Defines the type of storage location for the thread.
    """
    IN_MEMORY_MESSAGES = "InMemoryMessages"
    CONVERSATION_ID = "ConversationId"

class ChatClientAgentThread(AgentThread, MessagesRetrievableThread):
    """
    Chat client agent thread.
    """
    
    def __init__(self):
        """
        Initialize a new instance of the ChatClientAgentThread class.
        """
        super().__init__()
        self._messages: List[ChatMessage] = []
        self._storage_location: Optional[str] = None
        
    @property
    def storage_location(self) -> Optional[str]:
        """
        Gets the storage location for the thread.
        
        Returns:
            Optional[str]: The storage location or None.
        """
        return self._storage_location
        
    @storage_location.setter
    def storage_location(self, value: Optional[str]) -> None:
        """
        Sets the storage location for the thread.
        
        Args:
            value (Optional[str]): The storage location or None.
        """
        self._storage_location = value
        
    async def on_new_messages_async(self, messages, cancellation_token=None) -> None:
        """
        Called when new messages are added to the thread.
        Stores the messages if the thread is using InMemoryMessages storage.
        
        Args:
            messages: The new messages to add.
            cancellation_token: The token to monitor for cancellation requests.
        """
        if self.storage_location == ChatClientAgentThreadType.IN_MEMORY_MESSAGES:
            for message in messages:
                self._messages.append(message)
                
    async def get_messages_async(self, cancellation_token=None) -> AsyncIterator[ChatMessage]:
        """
        Gets the messages in the thread.
        For in-memory messages, returns the stored messages.
        For conversation ID, no messages are returned since they're stored on the server.
        
        Args:
            cancellation_token: The token to monitor for cancellation requests.
            
        Yields:
            ChatMessage: The messages in the thread.
        """
        if self.storage_location == ChatClientAgentThreadType.IN_MEMORY_MESSAGES:
            for message in self._messages:
                yield message
